File: 2016/grpc-stringdb/db_server.cpp
// Sample server code implementing the string database, using the gRPC service
// defined in stringdb.proto
//
// Eli Bendersky [http://eli.thegreenplace.net]
// This code is in the public domain.
#include <iostream>
#include <map>
#include <memory>
#include <mutex>
#include <string>

#include <grpc++/grpc++.h>
#include "stringdb.grpc.pb.h"

// Implementation of the StringDb service. We have to implement all of the
// service's methods - these will be invoked by the running gRPC server.
class StringDbImpl final : public stringdb::StringDb::Service {
public:
  grpc::Status GetValue(grpc::ServerContext* context,
                        const stringdb::GetValueRequest* request,
                        stringdb::GetValueReply* reply) override {
    std::lock_guard<std::mutex> lock(db_mutex_);

    auto it = string_db_.find(request->key());
    if (it == string_db_.end()) {
      reply->set_value("");
    } else {
      reply->set_value(it->second);
    }
    return grpc::Status::OK;
  }

  grpc::Status SetValue(grpc::ServerContext* context,
                        const stringdb::SetValueRequest* request,
                        stringdb::SetValueReply* reply) override {
    std::lock_guard<std::mutex> lock(db_mutex_);

    string_db_[request->key()] = request->value();
    reply->set_value(request->value());
    return grpc::Status::OK;
  }

  grpc::Status CountValue(grpc::ServerContext* context,
                          const stringdb::CountValueRequest* request,
                          stringdb::CountValueReply* reply) override {
    std::lock_guard<std::mutex> lock(db_mutex_);

    auto it = string_db_.find(request->key());
    if (it == string_db_.end()) {
      reply->set_count(-1);
    } else {
      reply->set_count(it->second.size());
    }
    return grpc::Status::OK;
  }

private:
  // The actual database.
  std::map<std::string, std::string> string_db_;

  // Mutex serializing access to the map.
  std::mutex db_mutex_;
};

void RunServer() {
  // This parts is taken from the "hello world" gRPC sample.
  std::string server_address("0.0.0.0:4050");
  StringDbImpl service;

  grpc::ServerBuilder builder;
  // Listen on the given address without any authentication mechanism.
  builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
  // Register "service" as the instance through which we'll communicate with
  // clients. In this case it corresponds to an *synchronous* service.
  builder.RegisterService(&service);
  // Finally assemble the server.
  std::unique_ptr<grpc::Server> server(builder.BuildAndStart());
  std::cout << "Server listening on " << server_address << std::endl;

  // Wait for the server to shutdown. Note that some other thread must be
  // responsible for shutting down the server for this call to ever return.
  server->Wait();
}

int main(int argc, char** argv) {
  RunServer();

  return 0;
}

File: 2016/grpc-stringdb/go/src/db_server.go
package main

import (
	"fmt"
	"net"
	pb "stringdb"
	"sync"

	"golang.org/x/net/context"
	"google.golang.org/grpc"
	"google.golang.org/grpc/grpclog"
)

type stringdbServer struct {
	mu sync.Mutex
	db map[string]string
}

func (s *stringdbServer) GetValue(ctx context.Context, r *pb.GetValueRequest) (*pb.GetValueReply, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	// Here we rely on the map's default of an empty string when no key exists.
	return &pb.GetValueReply{s.db[r.Key]}, nil
}

func (s *stringdbServer) SetValue(ctx context.Context, r *pb.SetValueRequest) (*pb.SetValueReply, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.db[r.Key] = r.Value
	return &pb.SetValueReply{r.Value}, nil
}

func (s *stringdbServer) CountValue(ctx context.Context, r *pb.CountValueRequest) (*pb.CountValueReply, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	retval := -1
	v, ok := s.db[r.Key]
	if ok {
		retval = len(v)
	}
	return &pb.CountValueReply{int64(retval)}, nil
}

func newServer() *stringdbServer {
	s := new(stringdbServer)
	s.db = make(map[string]string)
	return s
}

func main() {
	portnum := 4050
	listenAddr := fmt.Sprintf(":%d", portnum)
	lis, err := net.Listen("tcp", listenAddr)
	if err != nil {
		grpclog.Fatalf("failed to listen: %v", err)
	}
	var opts []grpc.ServerOption
	grpcServer := grpc.NewServer(opts...)
	pb.RegisterStringDbServer(grpcServer, newServer())

	fmt.Println("Server listening on", portnum)
	grpcServer.Serve(lis)
}
